go/host/l1/publisher.go

```diff
@@ -0,0 +1,308 @@
+package l1
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	gethlog "github.com/ethereum/go-ethereum/log"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/host"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+	"github.com/obscuronet/go-obscuro/go/common/retry"
+	"github.com/obscuronet/go-obscuro/go/ethadapter"
+	"github.com/obscuronet/go-obscuro/go/ethadapter/mgmtcontractlib"
+	"github.com/obscuronet/go-obscuro/go/wallet"
+	"github.com/pkg/errors"
+)
+
+const (
+	// Attempts to broadcast the rollup transaction to the L1. Worst-case, equates to 7 seconds, plus time per request.
+	l1TxTriesRollup = 3
+	// Attempts to send secret initialisation, request or response transactions to the L1. Worst-case, equates to 63 seconds, plus time per request.
+	l1TxTriesSecret = 7
+
+	// todo - these values have to be configurable
+	maxWaitForL1Receipt       = 100 * time.Second
+	retryIntervalForL1Receipt = 10 * time.Second
+	maxWaitForSecretResponse  = 120 * time.Second
+)
+
+type Publisher struct {
+	hostData        host.Identity
+	hostWallet      wallet.Wallet // Wallet used to issue ethereum transactions
+	ethClient       ethadapter.EthClient
+	mgmtContractLib mgmtcontractlib.MgmtContractLib // Library to handle Management Contract lib operations
+
+	repository host.L1BlockRepository
+	logger     gethlog.Logger
+
+	running atomic.Bool
+}
+
+func NewL1Publisher(hostData host.Identity, hostWallet wallet.Wallet, client ethadapter.EthClient, mgmtContract mgmtcontractlib.MgmtContractLib, repository host.L1BlockRepository, logger gethlog.Logger) *Publisher {
```

Comment:
It seems like it's more idiomatic in Go to follow the "accept interfaces, return structs" approach. Whereas we tend towards the Interface/impl approach that we're used to from other languages in our code.

Like really the interface doesn't need to be public, it could be a private interface in the enclave-guardian package because that's the only thing that needs it atm.

I made it public when you asked on that other PR because I thought other services will need the same dependencies (and like you said it's nice to be able to see laid out in one place what the responsibilities of all the services are).

It's already sort of useful doing it this way though, as an example the enclave-guardian doesn't know the that l1-publisher is a Service (if it wants to mock it, then it wouldn't have to mock Start/Stop/HealthStatus).

Similarly, the host doesn't know that the Service that got registered is an l1-publisher. Keeping the interfaces segregated like that makes things easier to test in theory I think.

If something else comes along and just needs to be able to do one of the things that l1-publisher can do it can just make a small interface like `P2PPeersProvider` for the l1-publisher rather than depend on the whole thing.

To conclude the essay, 😅 I think we should consider private implementations, large interfaces and anything with "Impl" in the name to be potential code smells.